s3transfer/download.py

```diff
@@ -0,0 +1,294 @@
+# language governing permissions and limitations under the License.
+import functools
+import logging
+import os
+import socket
+import math
+
+from botocore.exceptions import IncompleteReadError
+from botocore.vendored.requests.packages.urllib3.exceptions import \
+    ReadTimeoutError
+
+from s3transfer.compat import SOCKET_ERROR
+from s3transfer.exceptions import RetriesExceededError
+from s3transfer.exceptions import QueueShutdownError
+from s3transfer.utils import random_file_extension
+from s3transfer.utils import get_callbacks
+from s3transfer.utils import StreamReaderProgress
+from s3transfer.utils import ShutdownQueue
+from s3transfer.tasks import Task
+from s3transfer.tasks import TaskSubmitter
+
+
+logger = logging.getLogger(__name__)
+
+SHUTDOWN_SENTINEL = object()
+S3_RETRYABLE_ERRORS = (
+    socket.timeout, SOCKET_ERROR, ReadTimeoutError, IncompleteReadError
+)
+
+
+class DownloadTaskSubmitter(TaskSubmitter):
+    def __init__(self, client, config, osutil, executor, io_executor):
+        super(DownloadTaskSubmitter, self).__init__(
+            client, config, osutil, executor)
+        self._io_executor = io_executor
+
+    def _submit(self, transfer_future, transfer_coordinator):
+        if transfer_future.meta.size is None:
+            # If a size was not provided figure out the size for the
+            # user.
+            response = self._client.head_object(
+                Bucket=transfer_future.meta.call_args.bucket,
+                Key=transfer_future.meta.call_args.key,
+                **transfer_future.meta.call_args.extra_args
+            )
+            transfer_future.meta.provide_transfer_size(
+                response['ContentLength'])
+
+        # Figure out a temporary filename for the file.
+        temp_filename = (
+            transfer_future.meta.call_args.fileobj + os.extsep +
+            random_file_extension()
+        )
+        # If it is greater than threshold do a ranged download, otherwise
+        # do a regular GetObject download.
+        if transfer_future.meta.size < self._config.multipart_threshold:
+            self._submit_download_request(
+                temp_filename, transfer_future, transfer_coordinator)
+        else:
+            self._submit_ranged_download_request(
+                temp_filename, transfer_future, transfer_coordinator)
+
+    def _submit_download_request(
+            self, temp_filename, transfer_future, transfer_coordinator):
+        call_args = transfer_future.meta.call_args
+
+        # Get the needed callbacks for the task
+        progress_callbacks = get_callbacks(transfer_future, 'progress')
+        done_callbacks = get_callbacks(transfer_future, 'done')
+
+        # Create the io queue that will be used for this download
+        io_queue = ShutdownQueue(self._config.max_io_queue_size)
+
+        # Start up the io task to write the downloads to disk as they
+        # come in.
+        self._submit_io_write_task(
```

Comment:
Yep that is correct. I can rename the ``IOWriterTask`` to make it more clear if that helps.